server/import_certify.go

```diff
@@ -0,0 +1,188 @@
+	akPub, err := tpm2.Unmarshal[tpm2.TPMTPublic](req.GetAkPub())
+	if err != nil {
+		return err
+	}
+
+	akName, err := tpm2.ObjectName(akPub)
+	if err != nil {
+		return err
+	}
+
+	signature, err := tpm2.Unmarshal[tpm2.TPMTSignature](req.GetSignature())
+	if err != nil {
+		return err
+	}
+
+	hmac, err := signature.Signature.HMAC()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyHMAC(secret, req.GetCertifyInfo(), hmac); err != nil {
+		return err
+	}
+
+	attest, err := tpm2.Unmarshal[tpm2.TPMSAttest](req.GetCertifyInfo())
+	if err != nil {
+		return err
+	}
+
+	if err := verifyAttest(attest); err != nil {
+		return err
+	}
+
+	certify, err := attest.Attested.Certify()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyCertify(akName, certify); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// generateRestrictedHMACKey writes a new hmac to the input parameter and produces the pub/priv tpm2 structures
+func generateRestrictedHMACKey(hmacKey []byte) (*tpm2.TPMTPublic, *tpm2.TPMTSensitive) {
+	// Generate the random obfuscation value and key
+	obfuscate := make([]byte, 32)
+	rand.Read(obfuscate)
+	rand.Read(hmacKey[:])
+
+	// Unique for a KEYEDHASH object is H_nameAlg(obfuscate | key)
+	// See Part 1, "Public Area Creation"
+	h := sha256.New()
+	h.Write(obfuscate)
+	h.Write(hmacKey[:])
+
+	pub := &tpm2.TPMTPublic{
+		Type:    tpm2.TPMAlgKeyedHash,
+		NameAlg: tpm2.TPMAlgSHA256,
+		ObjectAttributes: tpm2.TPMAObject{
+			UserWithAuth: true,
+			NoDA:         true,
+			Restricted:   true,
+			SignEncrypt:  true,
+		},
+		Parameters: tpm2.NewTPMUPublicParms(tpm2.TPMAlgKeyedHash, &tpm2.TPMSKeyedHashParms{
+			Scheme: tpm2.TPMTKeyedHashScheme{
+				Scheme: tpm2.TPMAlgHMAC,
+				Details: tpm2.NewTPMUSchemeKeyedHash(tpm2.TPMAlgHMAC, &tpm2.TPMSSchemeHMAC{
+					HashAlg: tpm2.TPMAlgSHA256,
+				}),
+			},
+		}),
+		Unique: tpm2.NewTPMUPublicID(tpm2.TPMAlgKeyedHash, &tpm2.TPM2BDigest{
+			Buffer: h.Sum(nil),
+		}),
+	}
+
+	priv := &tpm2.TPMTSensitive{
+		SensitiveType: tpm2.TPMAlgKeyedHash,
+		SeedValue: tpm2.TPM2BDigest{
+			Buffer: obfuscate,
+		},
+		Sensitive: tpm2.NewTPMUSensitiveComposite(tpm2.TPMAlgKeyedHash, &tpm2.TPM2BSensitiveData{
+			Buffer: hmacKey[:],
+		}),
+	}
+
+	return pub, priv
+}
+
+// verifyHMAC checks the MAC on the given message.
+func verifyHMAC(hmacKey []byte, message []byte, ha *tpm2.TPMTHA) error {
+	if ha.HashAlg != tpm2.TPMAlgSHA256 {
+		return fmt.Errorf("%w %v (expected SHA256)", errWrongHashAlg, ha.HashAlg)
+	}
+	digest := sha256.Sum256(message)
+	h := hmac.New(sha256.New, hmacKey[:])
+	h.Write(digest[:])
+	if subtle.ConstantTimeCompare(ha.Digest, h.Sum(nil)) != 1 {
+		return errInvalidHMAC
+	}
+	return nil
+}
+
+// verifyAttest checks that the attestation structure has valid data
+func verifyAttest(attest *tpm2.TPMSAttest) error {
+	if attest.Magic != tpm2.TPMGeneratedValue {
+		return fmt.Errorf("%w: unexpected prefix %0x", errInvalidAttestation, attest.Magic)
+	}
+
+	// TODO: check qualified signer?
```

Comment:
I don't think we need to check the qualified signer, since the signer must be the same HMAC key we put into the challenge blob in order for the HMAC verify to pass